src/sync/app.hpp

```diff
@@ -58,6 +57,181 @@ class App {
+        std::string name;
+
+        // Whether or not the key is disabled.
+        bool disabled;
+    };
+
+    /**
+     * A client for the user API key authentication provider which
+     * can be used to create and modify user API keys. This
+     * client should only be used by an authenticated user.
+    */
+    class UserAPIKeyProviderClient {
+    public:
+        /**
+         * Creates a user API key that can be used to authenticate as the current user.
+         *
+         * - parameters:
+         *     - name: The name of the API key to be created.
+         *     - completion_block: A callback to be invoked once the call is complete.
+        */
+        void create_api_key(const std::string& name,
+                            std::function<void(std::shared_ptr<UserAPIKey>, Optional<AppError>)> completion_block);
+
+        /**
+         * Fetches a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to fetch.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void fetch_api_key(const realm::ObjectId& id,
+                           std::function<void(std::shared_ptr<UserAPIKey>, Optional<AppError>)> completion_block);
+
+        /**
+         * Fetches the user API keys associated with the current user.
+         *
+         * - parameters:
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void fetch_api_keys(std::function<void(std::vector<UserAPIKey>, Optional<AppError>)> completion_block);
+
+        /**
+         * Deletes a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to delete.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void delete_api_key(const realm::ObjectId& id,
+                            std::function<void(Optional<AppError>)> completion_block);
+
+        /**
+         * Enables a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to enable.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void enable_api_key(const realm::ObjectId& id,
+                            std::function<void(Optional<AppError>)> completion_block);
+
+        /**
+         * Disables a user API key associated with the current user.
+         *
+         * - parameters:
+         *     - id: The id of the API key to disable.
+         *     - completion_block: A callback to be invoked once the call is complete.
+         */
+        void disable_api_key(const realm::ObjectId& id,
+                             std::function<void(Optional<AppError>)> completion_block);
+    private:
+        UserAPIKeyProviderClient(std::weak_ptr<App>);
```

Comment:
Why use a weak_ptr?

Reply:
Doesn't have to be a weak_ptr in the final implementation, but the constructor will need to derive route/path information from the outer app.